enhancements/network/on-prem-service-load-balancers.md

```diff
@@ -0,0 +1,370 @@
+larger scale.
+
+While the BGP option is attractive for scaling reasons, it’s also more
+complicated and will not work in all environments.  It won’t work in an
+environment that won’t allow BGP advertisements from the cluster Nodes.  If a
+cluster uses a Network addon that also makes use of BGP, MetalLB integration
+will be more challenging.  For example see the [MetalLB page about Calico
+support](https://metallb.universe.tf/configuration/calico/).
+
+The layer 2 mode has the advantage of working in more environments.  We could
+also consider a MetalLB enhancement that makes it understand different L2
+domains and manage different IP address pools for each domain where SLBs may
+reside.
+
+### User Stories
+
+#### Story 1 - Easy Use with a Small Cluster
+
+As an administrator of a small cluster that resides entirely on a single layer
+2 domain, I would like to configure one or more ranges of IP addresses from my
+network that the cluster is free to use for Service Load Balancers.  I do not
+want to do any extra configuration in my network infrastructure beyond just
+ensuring that the configured ranges of addresses are not used elsewhere.
+
+MetalLB can do this today.
+
+#### Story 2 - BGP Integration
+
+As an administrator of a larger cluster with Nodes that reside on multiple L2
+network segments, I would like to configure one or more ranges of IP addresses
+from my network that the cluster is free to use for Service Load Balancers.  I
+would like my cluster to peer with my BGP infrastructure to advertise the
+current location of IP addresses allocated to Service Load Balancers.
+
+MetalLB can do this today.
+
+#### Story 3 - Larger Clusters without BGP
+
+As an administrator of a larger cluster with Nodes that reside on multiple L2
+network segments, I would like to configure one or more ranges of IP addresses
+from my network that the cluster is free to use for Service Load Balancers.  I
+do not have BGP infrastructure available or I'm not willing to have my cluster
+peer with my BGP infrastructure.  I would like to configure awareness of which
+subsets of my Nodes can use which pools of IP addresses since not every Node
+has physical connectivity to the same L2 networks.
+
+Note that MetalLB does not offer this today.  There has been some discussion
+about related functionality:
+* https://github.com/metallb/metallb/issues/605
+* https://github.com/metallb/metallb/pull/502
+
+### Implementation Details/Notes/Constraints
+
+The first step of implementation is to invest in the upstream project.  We
+should have one or more engineers engage with the project to handle issues,
+review pull requests, and contribute bug fixes or enhancements.  As part of
+this process we should continue our technical due diligence with testing and
+reviewing code to increase our confidence in choosing this solution.  One area
+we could contribute immediately is with setting up upstream CI.  The project
+does not appear to run any CI today.
```

Comment:
Based on our experiences in ovn-kubernetes upstream, I think [kind](https://github.com/kubernetes-sigs/kind/) is a good basis for CI of community kubernetes-ecosystem projects. It's not nearly as fancy as what we do in kubernetes and openshift, but you can use it with the built-in free github CI support rather than needing someone to be paying for an AWS/GCP account, etc.

(It would even allow testing both L2 and L3 mode. It doesn't matter what protocols the actual underlying network supports if you're doing all of your testing in a virtual network built on top of it.)

Reply:
Awesome, thanks for the feedback!  I'll incorporate this into the document.